File: python/set.py
#!/usr/bin/python

class Set(list):
    def __init__(self, value = []):
        list.__init__([])
        self.concat(value)

    def intersect(self, other):
        res = []
        for x in self:
            if x in other:
                res.append(x)
        return Set(res)

    def union(self, other):
        res = Set(self)
        res.concat(other)
        return res

    def concat(self, value):
        for x in value:
            if not x in self:
                self.append(x)

    def __and__(self, other):
        return self.intersect(other)
    def __or__(self, other):
        return self.union(other)
    def __repr__(self):
        return 'Set:' + list.__repr__(self)

File: ocaml/set.ml
let rec mem a = function
    [] -> false
  | (h::t) -> if a = h then true else mem a t

let rec intersect s = function
    [] -> []
  | (h::t) -> 
      if mem h s then 
	h :: intersect s t
      else intersect s t

let rec union s = function
    [] -> s
  | (h::t) -> 
      let res = union s t 
      in if mem h res then res else h :: res

let rec diff s = 
    let rec filter f = function
	[] -> []
      | (h::t) -> if f h then h::(filter f t) else filter f t in 
    function
	[] -> s
      | (lh::lt) -> 
	  if mem lh s then 
	    diff (filter ((<>) lh) s) lt
	  else
	    lh :: diff s lt
